calyx-opt/src/analysis/static_par_domination.rs

```diff
@@ -0,0 +1,313 @@
+                    // sort vec1 to get deterministic ordering
+                    vec1.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+                    for (enable_id, interval) in vec1 {
+                        // print enable id -- (interval)
+                        write!(f, "{:?} -- ", enable_id)?;
+                        writeln!(f, "{:?}", interval)?;
+                    }
+                }
+            }
+            write!(f, "====")?;
+            write!(f, "MAY EXECUTE")?;
+            writeln!(f, "====")?;
+            // print the "may executes" for the given par id
+            match node_may_map.get(par_id) {
+                None => (),
+                Some(map) => {
+                    let mut vec1: Vec<_> = map.iter().collect();
+                    // sort vec1 to get deterministic ordering
+                    vec1.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));
+                    for (enable_id, interval) in vec1 {
+                        // print enable id -- (interval)
+                        write!(f, "{:?} -- ", enable_id)?;
+                        writeln!(f, "{:?}", interval)?;
+                    }
+                }
+            }
+        }
+        writeln!(f)
+    }
+}
+
+impl StaticParDomination {
+    /// Construct a live range analysis.
+    pub fn new(control: &mut ir::Control, comp_name: ir::Id) -> Self {
+        let mut timing_map = StaticParDomination {
+            component_name: comp_name,
+            ..Default::default()
+        };
+        // compute_unique_ids is deterministic
+        // so if we're calling it after we've called live range analysis (and the
+        // control program hasn't changed, then this is unnecessary)
+        ControlId::compute_unique_ids(control, 0, true);
+
+        timing_map.build_time_map(control);
+
+        timing_map
+    }
+
+    /// returns a HashMap that maps node ids -> set of nodes that dominate it
+    /// It will only return the node ids that are dominated within the same
+    /// static par block, not all the dominators for the entire control program
+    pub fn get_static_dominators(&mut self) -> HashMap<u64, HashSet<u64>> {
+        let mut static_dom_map: HashMap<u64, HashSet<u64>> = HashMap::new();
+
+        for (par_id, node_interval_mapping) in &self.node_timing_map {
+            let empty_map = HashMap::new();
+            // these are the nodes that *may* execute for the given par id
+            let may_execute_enables =
+                match self.node_maybe_timing_map.get(par_id) {
+                    Some(mapping) => mapping,
+                    None => &empty_map,
+                };
+            // Very simple/naive algorithm
+            // for each "must" execute nodes, it can either dominate:
+            // one of the "may" execute nodes, or dominate another "must" execute node
+            // "may" execute nodes cannot dominate anybody
+            for (node_id1, (_, end1)) in node_interval_mapping.clone() {
```

Comment:
Why is this `clone()` necessary? Can you iterate over the set and only clone the underlying numbers when needed?